enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,851 @@
+set of CPUs, while other workloads will be prevented from running
+there. For the management workload case, this effectively gives 3
+pools of CPUs
+
+* **management** -- a user-defined cpuset that restricts where
+  management components run
+* **dedicated** -- CPUs used for exclusive workloads
+* **shared** -- CPUs in the set `management + !dedicate`
+
+Because the management CPU pool is restricted to management workloads,
+the shared CPU pool will need other CPUs to support burstable or
+best-effort workloads that are not identified as management workloads.
+
+We want the isolation to be in place from the initial buildout of the
+cluster, to ensure predictable behavior and to ensure that the cluster
+can fully operate with workload partitioning. Therefore, the feature
+must be enabled during installation. To enable the feature, the user
+will specify the set of CPUs to run all management workloads. When the
+management CPU set is not defined, the feature will be completely
+disabled.
+
+We generally want components to opt-in to workload partitioning and
+especially to being considered management workloads. Therefore, for a
+regular pod to be considered to contain a management workload it must
+be labeled with an annotation configuring its *workload type*,
+`workload.openshift.io/{workload-type}`. For now, we will focus on
+management workloads via `workload.openshift.io/management`, but will
+use a syntax that supports other types of workloads that may be
+defined in future enhancements.
+
+We want to be able to differentiate between "hard" and "soft"
+partitioning requests for workload types. We also want to be able to
+support other partitioning parameters in the future. Therefore,
+workload type settings will use an annotation with a JSON value (see
+below for details).
+
+We want to treat all OpenShift components as management workloads,
+including those that run the control plane. Therefore, kubelet will be
+modified to partition static pods with the annotation based on their
+workload type, when the feature is enabled. We will update operator
+manifests and implementations to annotate all OpenShift components not
+running in static pods has having the workload type "`management`",
+and add a CI job to require that the annotation be present in all
+workloads created from the release payload.
+
+We need kubelet to know when the feature is enabled, but we cannot
+change the configuration schema that kubelet inherits from
+upstream. Therefore we will have kubelet look for a new configuration
+file on startup, and the feature will only be enabled if that file is
+found.
+
+We want to give cluster administrators control over which workloads
+are run on the management CPUs. Normal users cannot add a label to a
+namespace without the right RBAC permissions. Therefore, only pods in
+namespaces labeled with a label
+`workload.openshift.io/allowed={comma_separated_list_of_type_names}`
```

Comment:
I think this is not an annotation? `workload.openshift.io/allowed: "management"`

Reply:
Line 359 says it's an annotation, not a label.